netplugin/netd.go

```diff
@@ -14,248 +14,295 @@ limitations under the License.
+		logrus.Infof("Set netplugin use syslog config: false")
 	}
 
-	if opts.debug {
-		log.SetLevel(log.DebugLevel)
-		os.Setenv("CONTIV_TRACE", "1")
+	logLevel, err := logrus.ParseLevel(ctx.String("log-level"))
+	if err != nil {
+		return plugin.Config{}, err
 	}
+	logrus.SetLevel(logLevel)
+	logrus.Infof("Use netplugin log level: %v", logLevel)
 
-	if opts.jsonLog {
-		log.SetFormatter(&log.JSONFormatter{})
+	if ctx.Bool("use-json-log") {
+		logrus.SetFormatter(&logrus.JSONFormatter{})
+		logrus.Infof("Use netplugin log format: json")
 	} else {
-		log.SetFormatter(&log.TextFormatter{FullTimestamp: true, TimestampFormat: time.StampNano})
+		logrus.SetFormatter(&logrus.TextFormatter{FullTimestamp: true, TimestampFormat: time.StampNano})
+		logrus.Infof("Use netplugin log format: text")
 	}
 
-	if opts.syslog != "" {
-		configureSyslog(opts.syslog)
+	// 2. validate and set plugin mode
+	pluginMode := strings.ToLower(ctx.String("mode"))
+	switch pluginMode {
+	case master.Docker, master.Kubernetes, master.SwarmMode, master.Test:
+		logrus.Infof("Use netplugin mode: %v", pluginMode)
+	case "":
+		return plugin.Config{}, fmt.Errorf("netplugin mode is not set")
+	default:
+		return plugin.Config{}, fmt.Errorf("Unknown netplugin mode: %v", pluginMode)
 	}
 
-	if flagSet.NFlag() < 1 {
-		log.Infof("host-label not specified, using default (%s)", opts.hostLabel)
+	// 3. validate and set network mode
+	networkMode := strings.ToLower(ctx.String("netmode"))
+	switch networkMode {
+	case "vlan", "vxlan":
+		logrus.Infof("Use netplugin network mode: %v", networkMode)
+	case "":
+		return plugin.Config{}, fmt.Errorf("netplugin network mode is not set")
+	default:
+		return plugin.Config{}, fmt.Errorf("Unknown netplugin network mode: %v", networkMode)
 	}
 
-	// default to using local IP addr
-	localIP, err := cluster.GetLocalAddr()
-	if err != nil {
-		log.Fatalf("Error getting local address. Err: %v", err)
+	// 4. validate forward mode
+	forwardMode := strings.ToLower(ctx.String("fwdmode"))
+	if (networkMode == "vlan" && forwardMode == "bridge") || forwardMode == "routing" {
+		logrus.Infof("Use netplugin forwarding mode: %v", forwardMode)
+	} else if forwardMode == "" {
+		return plugin.Config{}, fmt.Errorf("netplugin network mode is not set")
+	} else {
+		return plugin.Config{}, fmt.Errorf("Invalid netplugin forwarding mode: %v", forwardMode)
 	}
-	if opts.ctrlIP == "" {
-		opts.ctrlIP = localIP
+
+	// 5. validate and set other optional configs
+	hostLabel := ctx.String("host")
+	logrus.Infof("Use netplugin host: %v", hostLabel)
+	controlIP := ctx.String("ctrl-ip")
+	logrus.Infof("Use netplugin control IP: %v", controlIP)
+	vtepIP := ctx.String("vtep-ip")
+	if networkMode == "vxlan" && vtepIP == "" {
+		return plugin.Config{}, fmt.Errorf("vtep-ip should be set when using VXLAN mode")
 	}
-	if opts.vtepIP == "" {
-		opts.vtepIP = opts.ctrlIP
+	logrus.Infof("Use netplugin VTEP IP: %v", vtepIP)
+	vlanUpLinks := strings.Split(ctx.String("vlan-uplinks"), ",")
```

Comment:
this seems new? missing from commit msg? am I just tired?

Reply:
this is new (sort of). latest commit changed it from StringSlice type to Sting tyle

Reply:
k